### ISSUE TITLE
seekBefore is not fetching the previous but the first page

### ISSUE THREAD

**Issue by whiskeysierra** [jOOQ/jOOQ#6380]
According to #2805:

> The SEEK method is currently only partially implemented, i.e. it allows only for fetching the "next" page. In order to fetch the "previous" page, a similar seekBefore() method could be introduced.

But it what it generates is:

```sql
  SELECT * FROM table
   WHERE id < 5
ORDER BY id ASC
   LIMIT 3
```

compared to `seekAfter`:

```sql
  SELECT * FROM table
   WHERE id > 5
ORDER BY id ASC
   LIMIT 3
```

As you can see it only inverts the condition from `>` to `<`.
Let's assume that the table contains rows with id from 1 to 10.
Using `seekAfter(5).limit(3)` will return `6`, `7` and `8`,
but `seekBefore(5).limit(3)` will return `1`, `2` and `3`.
I think it really **should return** `2`, `3` and `4`.

I'd expect something like:

```sql
SELECT * FROM (
      SELECT * FROM table
       WHERE id < 5
    ORDER BY id DESC
       LIMIT 3
) AS t
ORDER BY id ASC
```

See also http://ask.use-the-index-luke.com/questions/205/how-to-query-for-previous-page-with-keyset-pagination

**Comment by lukaseder** [comment]
Thank you very much for reporting. We'll look into this ASAP

**Comment by lukaseder** [comment]
You're right, the expected behaviour is the one you're describing, and the one that was described in the original issue #2805. This slipped through because the integration tests for `seekBefore()` paginated back from page 2 to page 1, or they paginated without `LIMIT`, in case of which the current behaviour is accidentally correct.

Even if this is a behaviourally backwards-incompatible change, I'll assume that no one would want the current behaviour. Also, the change does not violate the Javadoc, but extend it, as the Javadoc currently doesn't make an example using `LIMIT`.

We'll fix this in jOOQ 3.10. Given the backwards incompatibility, the fix won't be merged to 3.9.4 or older releases.

**Comment by whiskeysierra** [comment]
For anyone who has the issue right now, there is a workaround if you need the behaviour of `seekBefore` as it is specified:
- flip all sort fields in your `orderBy` clause, e.g. `A.asc(), B.desc()` should become `A.desc(), B.asc()`
- use `seekAfter` instead of `seekBefore`
- reverse the result list using `Collections.reverse`

**Comment by lukaseder** [comment]
I'm afraid, this won't fit in jOOQ 3.10 anymore. There are some delicate details related to applying this with queries that involve unions to get right...

**Comment by whiskeysierra** [comment]
I think it's OK given that a workaround exists.

**Comment by lukaseder** [comment]
Looking into fixing this now. Things to consider:

- Derived tables are not allowed to have ambiguous column names, so we need to rename them to enumerated names and then rename the enumerated names back to the original ones in the outer query. There's already a mechanism in place for this to emulate offset pagination in older SQL dialects
- Speaking of emulated pagination (Oracle 11g and before), that has to work as well with both `SEEK BEFORE` and `SEEK AFTER`. Other databases don't have this problem as they all have had a way of doing `LIMIT` / `TOP` / `FETCH` for ages. Perhaps, we won't fix this.
- Set operations like `UNION` still don't work correctly with `SEEK BEFORE` and `SEEK AFTER`: #7459

Since there are quite some internals that might need refactoring, it makes sense to look at all of these problems in one go.

**Comment by lukaseder** [comment]
This is a really challenging change. Unfortunately, these things haven't been considered properly, prior to implementing this feature. It seems that deprecating it might be the best way forward.

The workaround you've documented is one option for simple cases. Another option is to manually wrap a `SEEK` query in a derived table and order it again.

**Comment by lukaseder** [comment]
Deprecation: #7461. it will be lifted again, once this is fixed.

**Comment by cowwoc** [comment]
> Another option is to manually wrap a SEEK query in a derived table and order it again.

How can we implement seekBack in terms of a derived table without knowing the column names? I want to implement seekBack once in a generic manner instead of having to reimplement once per query but then the implementation can't know the table, column or ordering that needs to be reversed ahead of time. Any ideas?

**Comment by lukaseder** [comment]
> How can we implement seekBack in terms of a derived table without knowing the column names?

You can rename the columns in a derived table by using derived column lists. SQL syntax: `(...) as t (a, b, c)`. jOOQ syntax: `Table.as(String, String...)`. This may be necessary anyway, in case you have duplicate column names in your derived table (allowed in top level selects, but not in subqueries)

Alternatively, when you don't rename the columns, you can still access them by using `Table.fields()`. Hope this helps.

**Comment by martinlau** [comment]
>  I'll assume that no one would want the current behaviour.

The current behaviour is actually useful in one situation. If you consider the combinations of first/last N with an optional before/after X:

1. First N = `.orderBy(X.asc(), Y.asc()).limit(n)`
2. Last N = `.orderBy(X.desc(), Y.desc()).limit(N).reversed()`
3. First N After ... = `.orderBy(X.asc(), Y.asc()).seekAfter(x, y).limit(n)`
4. Last N Before ... = `.orderBy(X.desc(), Y.desc()).seekAfter(x, y).limit(n).reversed()`
5. First N Before ... = `.orderBy(X.asc(), Y.asc()).seekBefore(x, y).limit(n)`
6. Last N After ... = `.orderBy(X.desc(), Y.desc()).seekBefore(x, y).limit(n).reversed()` 

No 1-4 are pretty unexciting and work with the current implementation.

No 5. Is broken with the current implementation. The solution (as per [whiskeysierra's comment](https://github.com/jOOQ/jOOQ/issues/6380#issuecomment-314779197)) is to use `.orderBy(X, Y, Z).seekAfter(flip(x), flip(y), flip(z)).limit(n).reverse()`.

No. 6 only works with the current implementation. The item passed to seekBefore is more like a limit or upper bound than a starting point (in the context of the initial report, I actually want 1, 2 and 3  - not 2, 3 and 4).

I'm not sure if there's a similar "flip it and reverse it" work-around once a fix is implemented? If not, it may be useful to keep an appropriately named version of the current implementation around.
